// react-native-reanimated/lib/commonjs/reanimated2/layoutReanimation/defaultTransitions/JumpingTransition.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.JumpingTransition = void 0;

var _animation = require("../../animation");

var _Easing = require("../../Easing");

var _animationBuilder = require("../animationBuilder");

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class JumpingTransition extends _animationBuilder.BaseAnimationBuilder {
  constructor() {
    super(...arguments);

    _defineProperty(this, "build", () => {
      const delayFunction = this.getDelayFunction();
      const callback = this.callbackV;
      const delay = this.getDelay();
      const duration = (this.durationV ?? 300) / 2;
      const config = {
        duration: duration * 2
      };
      return values => {
        'worklet';

        const d = Math.max(Math.abs(values.targetOriginX - values.currentOriginX), Math.abs(values.targetOriginY - values.currentOriginY));
        return {
          initialValues: {
            originX: values.currentOriginX,
            originY: values.currentOriginY,
            width: values.currentWidth,
            height: values.currentHeight
          },
          animations: {
            originX: delayFunction(delay, (0, _animation.withTiming)(values.targetOriginX, config)),
            originY: delayFunction(delay, (0, _animation.withSequence)((0, _animation.withTiming)(Math.min(values.targetOriginY, values.currentOriginY) - d, {
              duration,
              easing: _Easing.Easing.out(_Easing.Easing.exp)
            }), (0, _animation.withTiming)(values.targetOriginY, { ...config,
              duration,
              easing: _Easing.Easing.bounce
            }))),
            width: delayFunction(delay, (0, _animation.withTiming)(values.targetWidth, config)),
            height: delayFunction(delay, (0, _animation.withTiming)(values.targetHeight, config))
          },
          callback: callback
        };
      };
    });
  }

  static createInstance() {
    return new JumpingTransition();
  }

}

exports.JumpingTransition = JumpingTransition;
