// react-native-reanimated/lib/commonjs/reanimated2/layoutReanimation/defaultAnimations/Bounce.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BounceOutUp = exports.BounceOutRight = exports.BounceOutLeft = exports.BounceOutDown = exports.BounceOut = exports.BounceInUp = exports.BounceInRight = exports.BounceInLeft = exports.BounceInDown = exports.BounceIn = void 0;

var _animation = require("../../animation");

var _ComplexAnimationBuilder = require("../animationBuilder/ComplexAnimationBuilder");

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class BounceIn extends _ComplexAnimationBuilder.ComplexAnimationBuilder {
  constructor() {
    super(...arguments);

    _defineProperty(this, "build", () => {
      const delayFunction = this.getDelayFunction();
      const delay = this.getDelay();
      const duration = this.getDuration();
      const callback = this.callbackV;
      const initialValues = this.initialValues;
      return () => {
        'worklet';

        return {
          animations: {
            transform: [{
              scale: delayFunction(delay, (0, _animation.withSequence)((0, _animation.withTiming)(1.2, {
                duration: duration * 0.55
              }), (0, _animation.withTiming)(0.9, {
                duration: duration * 0.15
              }), (0, _animation.withTiming)(1.1, {
                duration: duration * 0.15
              }), (0, _animation.withTiming)(1, {
                duration: duration * 0.15
              })))
            }]
          },
          initialValues: {
            transform: [{
              scale: 0
            }],
            ...initialValues
          },
          callback: callback
        };
      };
    });
  }

  static createInstance() {
    return new BounceIn();
  }

  static getDuration() {
    return 600;
  }

  getDuration() {
    return this.durationV ?? 600;
  }

}

exports.BounceIn = BounceIn;

class BounceInDown extends _ComplexAnimationBuilder.ComplexAnimationBuilder {
  constructor() {
    super(...arguments);

    _defineProperty(this, "build", () => {
      const delayFunction = this.getDelayFunction();
      const delay = this.getDelay();
      const duration = this.getDuration();
      const callback = this.callbackV;
      const initialValues = this.initialValues;
      return values => {
        'worklet';

        return {
          animations: {
            transform: [{
              translateY: delayFunction(delay, (0, _animation.withSequence)((0, _animation.withTiming)(-20, {
                duration: duration * 0.55
              }), (0, _animation.withTiming)(10, {
                duration: duration * 0.15
              }), (0, _animation.withTiming)(-10, {
                duration: duration * 0.15
              }), (0, _animation.withTiming)(0, {
                duration: duration * 0.15
              })))
            }]
          },
          initialValues: {
            transform: [{
              translateY: values.windowHeight
            }],
            ...initialValues
          },
          callback: callback
        };
      };
    });
  }

  static createInstance() {
    return new BounceInDown();
  }

  static getDuration() {
    return 600;
  }

  getDuration() {
    return this.durationV ?? 600;
  }

}

exports.BounceInDown = BounceInDown;

class BounceInUp extends _ComplexAnimationBuilder.ComplexAnimationBuilder {
  constructor() {
    super(...arguments);

    _defineProperty(this, "build", () => {
      const delayFunction = this.getDelayFunction();
      const delay = this.getDelay();
      const duration = this.getDuration();
      const callback = this.callbackV;
      const initialValues = this.initialValues;
      return values => {
        'worklet';

        return {
          animations: {
            transform: [{
              translateY: delayFunction(delay, (0, _animation.withSequence)((0, _animation.withTiming)(20, {
                duration: duration * 0.55
              }), (0, _animation.withTiming)(-10, {
                duration: duration * 0.15
              }), (0, _animation.withTiming)(10, {
                duration: duration * 0.15
              }), (0, _animation.withTiming)(0, {
                duration: duration * 0.15
              })))
            }]
          },
          initialValues: {
            transform: [{
              translateY: -values.windowHeight
            }],
            ...initialValues
          },
          callback: callback
        };
      };
    });
  }

  static createInstance() {
    return new BounceInUp();
  }

  static getDuration() {
    return 600;
  }

  getDuration() {
    return this.durationV ?? 600;
  }

}

exports.BounceInUp = BounceInUp;

class BounceInLeft extends _ComplexAnimationBuilder.ComplexAnimationBuilder {
  constructor() {
    super(...arguments);

    _defineProperty(this, "build", () => {
      const delayFunction = this.getDelayFunction();
      const delay = this.getDelay();
      const duration = this.getDuration();
      const callback = this.callbackV;
      const initialValues = this.initialValues;
      return values => {
        'worklet';

        return {
          animations: {
            transform: [{
              translateX: delayFunction(delay, (0, _animation.withSequence)((0, _animation.withTiming)(20, {
                duration: duration * 0.55
              }), (0, _animation.withTiming)(-10, {
                duration: duration * 0.15
              }), (0, _animation.withTiming)(10, {
                duration: duration * 0.15
              }), (0, _animation.withTiming)(0, {
                duration: duration * 0.15
              })))
            }]
          },
          initialValues: {
            transform: [{
              translateX: -values.windowWidth
            }],
            ...initialValues
          },
          callback: callback
        };
      };
    });
  }

  static createInstance() {
    return new BounceInLeft();
  }

  static getDuration() {
    return 600;
  }

  getDuration() {
    return this.durationV ?? 600;
  }

}

exports.BounceInLeft = BounceInLeft;

class BounceInRight extends _ComplexAnimationBuilder.ComplexAnimationBuilder {
  constructor() {
    super(...arguments);

    _defineProperty(this, "build", () => {
      const delayFunction = this.getDelayFunction();
      const delay = this.getDelay();
      const duration = this.getDuration();
      const callback = this.callbackV;
      const initialValues = this.initialValues;
      return values => {
        'worklet';

        return {
          animations: {
            transform: [{
              translateX: delayFunction(delay, (0, _animation.withSequence)((0, _animation.withTiming)(-20, {
                duration: duration * 0.55
              }), (0, _animation.withTiming)(10, {
                duration: duration * 0.15
              }), (0, _animation.withTiming)(-10, {
                duration: duration * 0.15
              }), (0, _animation.withTiming)(0, {
                duration: duration * 0.15
              })))
            }]
          },
          initialValues: {
            transform: [{
              translateX: values.windowWidth
            }],
            ...initialValues
          },
          callback: callback
        };
      };
    });
  }

  static createInstance() {
    return new BounceInRight();
  }

  static getDuration() {
    return 600;
  }

  getDuration() {
    return this.durationV ?? 600;
  }

}

exports.BounceInRight = BounceInRight;

class BounceOut extends _ComplexAnimationBuilder.ComplexAnimationBuilder {
  constructor() {
    super(...arguments);

    _defineProperty(this, "build", () => {
      const delayFunction = this.getDelayFunction();
      const delay = this.getDelay();
      const duration = this.getDuration();
      const callback = this.callbackV;
      const initialValues = this.initialValues;
      return () => {
        'worklet';

        return {
          animations: {
            transform: [{
              scale: delayFunction(delay, (0, _animation.withSequence)((0, _animation.withTiming)(1.1, {
                duration: duration * 0.15
              }), (0, _animation.withTiming)(0.9, {
                duration: duration * 0.15
              }), (0, _animation.withTiming)(1.2, {
                duration: duration * 0.15
              }), (0, _animation.withTiming)(0, {
                duration: duration * 0.55
              })))
            }]
          },
          initialValues: {
            transform: [{
              scale: 1
            }],
            ...initialValues
          },
          callback: callback
        };
      };
    });
  }

  static createInstance() {
    return new BounceOut();
  }

  static getDuration() {
    return 600;
  }

  getDuration() {
    return this.durationV ?? 600;
  }

}

exports.BounceOut = BounceOut;

class BounceOutDown extends _ComplexAnimationBuilder.ComplexAnimationBuilder {
  constructor() {
    super(...arguments);

    _defineProperty(this, "build", () => {
      const delayFunction = this.getDelayFunction();
      const delay = this.getDelay();
      const duration = this.getDuration();
      const callback = this.callbackV;
      const initialValues = this.initialValues;
      return values => {
        'worklet';

        return {
          animations: {
            transform: [{
              translateY: delayFunction(delay, (0, _animation.withSequence)((0, _animation.withTiming)(-10, {
                duration: duration * 0.15
              }), (0, _animation.withTiming)(10, {
                duration: duration * 0.15
              }), (0, _animation.withTiming)(-20, {
                duration: duration * 0.15
              }), (0, _animation.withTiming)(values.windowHeight, {
                duration: duration * 0.55
              })))
            }]
          },
          initialValues: {
            transform: [{
              translateY: 0
            }],
            ...initialValues
          },
          callback: callback
        };
      };
    });
  }

  static createInstance() {
    return new BounceOutDown();
  }

  static getDuration() {
    return 600;
  }

  getDuration() {
    return this.durationV ?? 600;
  }

}

exports.BounceOutDown = BounceOutDown;

class BounceOutUp extends _ComplexAnimationBuilder.ComplexAnimationBuilder {
  constructor() {
    super(...arguments);

    _defineProperty(this, "build", () => {
      const delayFunction = this.getDelayFunction();
      const delay = this.getDelay();
      const duration = this.getDuration();
      const callback = this.callbackV;
      const initialValues = this.initialValues;
      return values => {
        'worklet';

        return {
          animations: {
            transform: [{
              translateY: delayFunction(delay, (0, _animation.withSequence)((0, _animation.withTiming)(10, {
                duration: duration * 0.15
              }), (0, _animation.withTiming)(-10, {
                duration: duration * 0.15
              }), (0, _animation.withTiming)(20, {
                duration: duration * 0.15
              }), (0, _animation.withTiming)(-values.windowHeight, {
                duration: duration * 0.55
              })))
            }]
          },
          initialValues: {
            transform: [{
              translateY: 0
            }],
            ...initialValues
          },
          callback: callback
        };
      };
    });
  }

  static createInstance() {
    return new BounceOutUp();
  }

  static getDuration() {
    return 600;
  }

  getDuration() {
    return this.durationV ?? 600;
  }

}

exports.BounceOutUp = BounceOutUp;

class BounceOutLeft extends _ComplexAnimationBuilder.ComplexAnimationBuilder {
  constructor() {
    super(...arguments);

    _defineProperty(this, "build", () => {
      const delayFunction = this.getDelayFunction();
      const delay = this.getDelay();
      const duration = this.getDuration();
      const callback = this.callbackV;
      const initialValues = this.initialValues;
      return values => {
        'worklet';

        return {
          animations: {
            transform: [{
              translateX: delayFunction(delay, (0, _animation.withSequence)((0, _animation.withTiming)(10, {
                duration: duration * 0.15
              }), (0, _animation.withTiming)(-10, {
                duration: duration * 0.15
              }), (0, _animation.withTiming)(20, {
                duration: duration * 0.15
              }), (0, _animation.withTiming)(-values.windowWidth, {
                duration: duration * 0.55
              })))
            }]
          },
          initialValues: {
            transform: [{
              translateX: 0
            }],
            ...initialValues
          },
          callback: callback
        };
      };
    });
  }

  static createInstance() {
    return new BounceOutLeft();
  }

  static getDuration() {
    return 600;
  }

  getDuration() {
    return this.durationV ?? 600;
  }

}

exports.BounceOutLeft = BounceOutLeft;

class BounceOutRight extends _ComplexAnimationBuilder.ComplexAnimationBuilder {
  constructor() {
    super(...arguments);

    _defineProperty(this, "build", () => {
      const delayFunction = this.getDelayFunction();
      const delay = this.getDelay();
      const duration = this.getDuration();
      const callback = this.callbackV;
      const initialValues = this.initialValues;
      return values => {
        'worklet';

        return {
          animations: {
            transform: [{
              translateX: delayFunction(delay, (0, _animation.withSequence)((0, _animation.withTiming)(-10, {
                duration: duration * 0.15
              }), (0, _animation.withTiming)(10, {
                duration: duration * 0.15
              }), (0, _animation.withTiming)(-20, {
                duration: duration * 0.15
              }), (0, _animation.withTiming)(values.windowWidth, {
                duration: duration * 0.55
              })))
            }]
          },
          initialValues: {
            transform: [{
              translateX: 0
            }],
            ...initialValues
          },
          callback: callback
        };
      };
    });
  }

  static createInstance() {
    return new BounceOutRight();
  }

  static getDuration() {
    return 600;
  }

  getDuration() {
    return this.durationV ?? 600;
  }

}

exports.BounceOutRight = BounceOutRight;
